### PR TITLE
Ro replica

### PR DESCRIPTION
**Pull request by toly-kournik**
1. Introduce ReadOnlyReplica class
    2. Introduce a new message type sent by ReadOnlyReplica - AskForCheckpointMsg and
       a new handler for it in ReplicaImp
    3. Small additional configuration changes in favour of ReadOnlyReplica

### CODE REVIEW

File: bftengine/src/bftengine/BFTEngine.cpp [hunk 1]
```diff
@@ -138,7 +139,7 @@ IReplica *IReplica::createNewReplica(ReplicaConfig *replicaConfig,
     ((PersistentStorageImp *)persistentStoragePtr.get())->init(move(metadataStoragePtr));
   }
 
-  auto *replicaInternal = new ReplicaInternal();
+  auto replicaInternal = new ReplicaInternal();
```
Comment (yuliasherman) [inline bftengine/src/bftengine/BFTEngine.cpp]: As for me, "auto *" helps to understand that an object of type pointer is created. I would leave it as is.

File: bftengine/src/bftengine/MsgsCommunicator.cpp [hunk 2]
```diff
@@ -17,12 +17,9 @@ namespace bftEngine::impl {
 using namespace std;
 
 MsgsCommunicator::MsgsCommunicator(ICommunication* comm,
-                                   shared_ptr<IncomingMsgsStorage>& incomingMsgsStorage,
-                                   shared_ptr<IReceiver>& msgReceiver) {
-  communication_ = comm;
-  incomingMsgsStorage_ = incomingMsgsStorage;
-  msgReceiver_ = msgReceiver;
-}
+                                   shared_ptr<IncomingMsgsStorage> incomingMsgsStorage,
```
Comment (yuliasherman) [inline bftengine/src/bftengine/MsgsCommunicator.cpp]: Was boring?

File: bftengine/src/bftengine/ReadOnlyReplica.cpp [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+    } else {
+      numRelevant++;
+      tableItrator++;
+    }
+  }
+  Assert(numRelevant == tableOfStableCheckpoints.size());
+  LOG_INFO(GL, "numRelevant=" << numRelevant);
+
+  // if enough - invoke state transfer
+  if (numRelevant >= config_.fVal + 1) {
```
Comment (yuliasherman) [inline bftengine/src/bftengine/ReadOnlyReplica.cpp]: A hardcoded condition? Would add a function that returns (config_.fVal + 1).

File: bftengine/src/bftengine/ReadOnlyReplica.hpp [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+class PersistentStorage;
+/**
+ *
+ */
+class ReadOnlyReplica : public ReplicaForStateTransfer {
+ public:
+  ReadOnlyReplica(const ReplicaConfig&,
+                  IStateTransfer*,
+                  std::shared_ptr<MsgsCommunicator>,
+                  std::shared_ptr<PersistentStorage> persistentStorage,
```
Comment (yuliasherman) [inline bftengine/src/bftengine/ReadOnlyReplica.hpp]: One of two: define all parameter names or remove all of them.

File: bftengine/src/bftengine/ReplicaImp.cpp [hunk 5]
```diff
@@ -1255,6 +1259,30 @@ void ReplicaImp::onMessage<CheckpointMsg>(CheckpointMsg *msg) {
...
+template <>
+void ReplicaImp::onMessage<AskForCheckpointMsg>(AskForCheckpointMsg *msg) {
+  // metric_received_checkpoints_.Get().Inc(); // TODO [TK]
+
+  LOG_INFO(GL, "Node " << config_.replicaId << " received AskForCheckpoint message from node " << msg->senderId());
+
+  const CheckpointInfo &checkpointInfo = checkpointsLog->get(lastStableSeqNum);
+  CheckpointMsg *checkpointMsg = checkpointInfo.selfCheckpointMsg();
+
+  if (checkpointMsg == nullptr) {
```
Comment (yuliasherman) [inline bftengine/src/bftengine/ReplicaImp.cpp]: Does nothing?

File: bftengine/src/bftengine/ReplicaImp.cpp [hunk 6]
```diff
@@ -2220,7 +2248,7 @@ void ReplicaImp::tryToSendReqMissingDataMsg(SeqNum seqNumber, bool slowPathOnly,
   const bool missingBigRequests = (!missingPrePrepare) && (!seqNumInfo.hasPrePrepareMsg());
 
   ReplicaId firstRepId = 0;
```
Comment (yuliasherman) [inline bftengine/src/bftengine/ReplicaImp.cpp]: Confusing name firstRepId - reply? replica? report?

File: bftengine/src/bftengine/ReplicasInfo.cpp [hunk 7]
```diff
@@ -10,41 +10,38 @@
...
+        return ret;
+      }()},
+
+      _idsOfPeerROReplicas{[config]() {
+        std::set<ReplicaId> ret;
+        for (auto i = config.numReplicas; i < config.numReplicas + config.numRoReplicas; ++i)
+          if (i != config.replicaId) ret.insert(i);
+        return ret;
+      }()} {
+  Assert(_numberOfReplicas == (3 * _fVal + 2 * _cVal + 1));
```
Comment (yuliasherman) [inline bftengine/src/bftengine/ReplicasInfo.cpp]: (3 * _fVal + 2 * _cVal + 1) : use config.numReplicas

File: bftengine/src/bftengine/messages/AskForCheckpointMsg.hpp [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+// This product may include a number of subcomponents with separate copyright notices and license terms. Your use of
+// these subcomponents is subject to the terms and conditions of the subcomponent's license, as noted in the LICENSE
+// file.
+
+#pragma once
+
+#include "MessageBase.hpp"
+#include "Digest.hpp"
+
+namespace bftEngine {
```
Comment (yuliasherman) [inline bftengine/src/bftengine/messages/AskForCheckpointMsg.hpp]: Combine workspaces

File: bftengine/src/bftengine/messages/AskForCheckpointMsg.hpp [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+
+ protected:
+#pragma pack(push, 1)
+  struct AskForCheckpointMsgHeader {
+    MessageBase::Header header;
+  };
+#pragma pack(pop)
+  static_assert(sizeof(AskForCheckpointMsgHeader) == sizeof(MessageBase::Header), "AskForCheckpointMsgHeader is 2B");
+
+  AskForCheckpointMsgHeader* b() const { return (AskForCheckpointMsgHeader*)msgBody_; }
```
Comment (yuliasherman) [inline bftengine/src/bftengine/messages/AskForCheckpointMsg.hpp]: b()? call it body()

File: bftengine/src/bftengine/ReplicasInfo.cpp (hunk 7 above)
Comment (toly-kournik) [reply bftengine/src/bftengine/ReplicasInfo.cpp]: _numberOfReplicas is initialized from config.numReplicas,
This is and existing assertion checking sanity

File: bftengine/src/bftengine/ReplicaImp.cpp [hunk 10]
```diff
@@ -1255,6 +1259,30 @@ void ReplicaImp::onMessage<CheckpointMsg>(CheckpointMsg *msg) {
     onReportAboutLateReplica(msgSenderId, msgSeqNum);
   }
 }
+/**
+ * Is sent from a read-only replica
+ */
+template <>
+void ReplicaImp::onMessage<AskForCheckpointMsg>(AskForCheckpointMsg *msg) {
```
Comment (yuliasherman) [inline bftengine/src/bftengine/ReplicaImp.cpp]: Don't you need to validate this message first?

Comment (toly-kournik) [reply bftengine/src/bftengine/ReplicaImp.cpp]: it's validated in ReplicaImp::messageHandler

File: bftengine/src/bftengine/ReplicaImp.cpp (hunk 5 above)
Comment (toly-kournik) [reply bftengine/src/bftengine/ReplicaImp.cpp]: I'm not sure yet if it's right to uncomment, I'll consult Guy and decide.

File: bftengine/src/bftengine/ReadOnlyReplica.cpp (hunk 3 above)
Comment (toly-kournik) [reply bftengine/src/bftengine/ReadOnlyReplica.cpp]: Copied the code from appropriate place in ReplicaImp

File: bftengine/src/bftengine/ReadOnlyReplica.hpp (hunk 4 above)
Comment (toly-kournik) [reply bftengine/src/bftengine/ReadOnlyReplica.hpp]: I'll do it in the next commit

File: bftengine/src/bftengine/messages/AskForCheckpointMsg.hpp (hunk 9 above)
Comment (toly-kournik) [reply bftengine/src/bftengine/messages/AskForCheckpointMsg.hpp]: There is already a body() method in the base class.
b() is a convention through over all the message classes

File: bftengine/src/bftengine/messages/AskForCheckpointMsg.hpp (hunk 8 above)
Comment (toly-kournik) [reply bftengine/src/bftengine/messages/AskForCheckpointMsg.hpp]: ok

File: bftengine/src/bftengine/ReplicaImp.cpp (hunk 6 above)
Comment (toly-kournik) [reply bftengine/src/bftengine/ReplicaImp.cpp]: I haven't touched this

File: bftengine/src/bftengine/ReadOnlyReplica.cpp (hunk 3 above)
Comment (salieri11) [reply bftengine/src/bftengine/ReadOnlyReplica.cpp]: i suggest to change to > and remove + 1. Much more intuitive

**Pull request merged**